service/service.go

```diff
@@ -115,43 +137,133 @@ func (s *Service) Start() {
 		if err != nil {
 			log.Fatalf("Error starting Service due to %v", err)
 		}
+		log.Debugf("server shut down")
 	}()
 
-	//wait for SIGTERM
+	s.prepareShutdown(h)
+}
+
+func (s *Service) prepareShutdown(h Server) {
+	s.Stop = make(chan os.Signal, 1)
+	signal.Notify(s.Stop, os.Interrupt, syscall.SIGTERM)
 	<-s.Stop
-	// we linebreak here just to get the log message printed nicely
-	fmt.Print("\n")
-	log.Warnf("Service shutting down...")
-	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
-	defer cancel() // Cancel ctx as soon as handleSearch returns.
-	//TODO: build some connection drainer for websockets
-	h.Shutdown(ctx)
-	log.Infof("Service stopped gracefully.")
+	shutdown(h)
 }
 
 // Shutdown allows to stop the HTTP Server gracefully
 func (s *Service) Shutdown() {
 	s.Stop <- os.Signal(os.Interrupt)
 }
 
+func shutdown(s Shutdowner) {
+	if s == nil {
+		return
+	}
+
+	// todo: make configurable
+	timeout := time.Second * 5
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	log.Printf("\nServer shutdown with timeout: %s\n", timeout)
+
+	if err := s.Shutdown(ctx); err != nil {
+		log.Printf("Error: %v", err)
+	} else {
+		if hs, ok := s.(*http.Server); ok {
+			log.Printf("Finished all in-flight HTTP requests\n")
+
+			if hss, ok := hs.Handler.(Shutdowner); ok {
+				select {
+				case <-ctx.Done():
+					if err := ctx.Err(); err != nil {
+						log.Printf("Error: %v", err)
+						return
+					}
+				default:
+					if deadline, ok := ctx.Deadline(); ok {
+						secs := (time.Until(deadline) + time.Second/2) / time.Second
+						log.Printf("Shutting down handler with timeout: %ds\n", secs)
+					}
+
+					done := make(chan error)
+
+					go func() {
+						<-ctx.Done()
+						done <- ctx.Err()
+					}()
+
+					go func() {
+						done <- hss.Shutdown(ctx)
+					}()
+
+					if err := <-done; err != nil {
+						log.Printf("Error: %v", err)
+						return
+					}
+				}
+			}
+		}
+
+		if deadline, ok := ctx.Deadline(); ok {
+			secs := (time.Until(deadline) + time.Second/2) / time.Second
+			log.Printf("Shutdown finished %ds before deadline\n", secs)
+		}
+	}
+}
+
 // prepareBeforeStart sets up the standard handlers
 func (s *Service) prepareBeforeStart() {
 	s.timer = NewTimer()
 	s.Router.Handle("/metrics", promhttp.Handler()).Methods("GET")
 	s.Router.HandleFunc("/health", s.healthHandler).Methods("GET")
 	s.Router.HandleFunc("/info", s.infoHandler).Methods("GET")
 
-	http.Handle("/", s.Router)
+	s.ServeMux.Handle("/", s.Router)
+
+	initPublicKey()
 }
 
 // HandleFunc excepts a HanderFunc an converts it to a handler, then registers this handler
+// Deprecated: Developers should use SecureHandleFunc() or UnsecureHandleFunc() explicitly
```

Comment:
Can we print a warning to console when used?

Reply:
I would not, spams you logs. No need to refactor now.